File: sdk/java/src/main/java/com/merchantplatform/Signing.java
package com.merchantplatform;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.SecureRandom;
import java.util.ArrayList;
import java.util.Base64;
import java.util.Comparator;
import java.util.List;
import java.util.Map;

public final class Signing {
    private Signing() {}
    public record SignedHeaders(String keyId, String timestamp, String nonce, String contentDigest, String signature) {
        public Map<String,String> asMap() { return Map.of("Merchant-Key-Id",keyId,"Merchant-Timestamp",timestamp,"Merchant-Nonce",nonce,"Content-Digest",contentDigest,"Merchant-Signature",signature); }
    }
    public static SignedHeaders signRequest(String keyId, String secret, String method, String pathAndQuery, byte[] body, long timestamp, String nonce) {
        try {
            byte[] digest=MessageDigest.getInstance("SHA-256").digest(body); String hex=hex(digest); String canonical=method.toUpperCase()+"\n"+pathAndQuery+"\n"+timestamp+"\n"+nonce+"\n"+hex;
            Mac mac=Mac.getInstance("HmacSHA256");mac.init(new SecretKeySpec(secret.getBytes(StandardCharsets.UTF_8),"HmacSHA256"));String signature=Base64.getUrlEncoder().withoutPadding().encodeToString(mac.doFinal(canonical.getBytes(StandardCharsets.UTF_8)));
            return new SignedHeaders(keyId,Long.toString(timestamp),nonce,"sha-256=:"+Base64.getEncoder().encodeToString(digest)+":",signature);
        } catch (Exception error) { throw new IllegalStateException("request signing unavailable", error); }
    }
    public static String canonicalQuery(Map<String,? extends List<?>> query) { List<String> pairs=new ArrayList<>();query.entrySet().stream().sorted(Comparator.comparing(Map.Entry::getKey)).forEach(entry->{for(Object value:entry.getValue())if(value!=null)pairs.add(formEncode(entry.getKey())+"="+formEncode(String.valueOf(value)));});return String.join("&",pairs); }
    public static String pathSegment(String value) { return percentEncode(value, false); }
    public static String randomNonce() { byte[] bytes=new byte[16];new SecureRandom().nextBytes(bytes);return hex(bytes); }
    public static String base64Url(byte[] value) { return Base64.getUrlEncoder().withoutPadding().encodeToString(value); }
    public static String digest(byte[] value) { try{return "sha-256=:"+Base64.getEncoder().encodeToString(MessageDigest.getInstance("SHA-256").digest(value))+":";}catch(Exception error){throw new IllegalStateException(error);} }
    private static String formEncode(String value) { return percentEncode(value, true); }
    private static String percentEncode(String value, boolean plusSpace) { StringBuilder result=new StringBuilder();for(byte item:value.getBytes(StandardCharsets.UTF_8)){int b=item&255;if((b>='a'&&b<='z')||(b>='A'&&b<='Z')||(b>='0'&&b<='9')||b=='-'||b=='.'||b=='_'||b=='~')result.append((char)b);else if(plusSpace&&b==' ')result.append('+');else result.append('%').append(String.format("%02X",b));}return result.toString(); }
    private static String hex(byte[] value) { StringBuilder result=new StringBuilder();for(byte item:value)result.append(String.format("%02x",item));return result.toString(); }
}

File: sdk/dotnet/src/MerchantPlatform/Signing.cs
using System.Security.Cryptography;
using System.Text;

namespace MerchantPlatform;

public sealed record SignedHeaders(string KeyId, string Timestamp, string Nonce, string ContentDigest, string Signature) {
    public IReadOnlyDictionary<string,string> AsDictionary() => new Dictionary<string,string>{{"Merchant-Key-Id",KeyId},{"Merchant-Timestamp",Timestamp},{"Merchant-Nonce",Nonce},{"Content-Digest",ContentDigest},{"Merchant-Signature",Signature}};
}
public static class Signing {
    public static SignedHeaders SignRequest(string keyId,string secret,string method,string pathAndQuery,ReadOnlySpan<byte> body,long timestamp,string nonce){var digest=SHA256.HashData(body);var canonical=string.Join("\n",method.ToUpperInvariant(),pathAndQuery,timestamp.ToString(System.Globalization.CultureInfo.InvariantCulture),nonce,Convert.ToHexString(digest).ToLowerInvariant());using var hmac=new HMACSHA256(Encoding.UTF8.GetBytes(secret));var signature=Base64Url(hmac.ComputeHash(Encoding.UTF8.GetBytes(canonical)));return new SignedHeaders(keyId,timestamp.ToString(System.Globalization.CultureInfo.InvariantCulture),nonce,$"sha-256=:{Convert.ToBase64String(digest)}:",signature);}
    public static string CanonicalQuery(IReadOnlyDictionary<string,IReadOnlyList<object?>> query){var pairs=new List<string>();foreach(var item in query.OrderBy(item=>item.Key,StringComparer.Ordinal))foreach(var value in item.Value)if(value is not null)pairs.Add(FormEncode(item.Key)+"="+FormEncode(Convert.ToString(value,System.Globalization.CultureInfo.InvariantCulture)!));return string.Join('&',pairs);}
    public static string RandomNonce()=>Convert.ToHexString(RandomNumberGenerator.GetBytes(16)).ToLowerInvariant();
    public static string PathSegment(string value)=>PercentEncode(value,false);
    public static string Digest(ReadOnlySpan<byte> value)=>$"sha-256=:{Convert.ToBase64String(SHA256.HashData(value))}:";
    public static string Base64Url(byte[] value)=>Convert.ToBase64String(value).TrimEnd('=').Replace('+','-').Replace('/','_');
    private static string FormEncode(string value)=>PercentEncode(value,true);
    private static string PercentEncode(string value,bool plusSpace){var result=new StringBuilder();foreach(var item in Encoding.UTF8.GetBytes(value)){var unreserved=(item>=(byte)'a'&&item<=(byte)'z')||(item>=(byte)'A'&&item<=(byte)'Z')||(item>=(byte)'0'&&item<=(byte)'9')||item==(byte)'-'||item==(byte)'.'||item==(byte)'_'||item==(byte)'~';if(unreserved)result.Append((char)item);else if(plusSpace&&item==(byte)' ')result.Append('+');else result.Append('%').Append(item.ToString("X2"));}return result.ToString();}
}
